subgraphs/vesta-finance/src/entities/market.ts

```diff
@@ -0,0 +1,375 @@
+  marketSnapshot.cumulativeLiquidateUSD = market.cumulativeLiquidateUSD;
+  marketSnapshot.inputTokenBalance = market.inputTokenBalance;
+  marketSnapshot.inputTokenPriceUSD = market.inputTokenPriceUSD;
+  marketSnapshot.outputTokenSupply = market.outputTokenSupply;
+  marketSnapshot.outputTokenPriceUSD = market.outputTokenPriceUSD;
+  marketSnapshot.exchangeRate = market.exchangeRate;
+  marketSnapshot.rewardTokenEmissionsAmount = market.rewardTokenEmissionsAmount;
+  marketSnapshot.rewardTokenEmissionsUSD = market.rewardTokenEmissionsUSD;
+  marketSnapshot.blockNumber = event.block.number;
+  marketSnapshot.timestamp = event.block.timestamp;
+  marketSnapshot.save();
+  return marketSnapshot;
+}
+
+// create seperate InterestRate Entities for each market snapshot
+// this is needed to prevent snapshot rates from being pointers to the current rate
+function getSnapshotRates(rates: string[], timeSuffix: string): string[] {
+  let snapshotRates: string[] = [];
+  for (let i = 0; i < rates.length; i++) {
+    let rate = InterestRate.load(rates[i]);
+    if (!rate) {
+      log.warning("[getSnapshotRates] rate {} not found, should not happen", [
+        rates[i],
+      ]);
+      continue;
+    }
+
+    // create new snapshot rate
+    let snapshotRateId = rates[i].concat("-").concat(timeSuffix);
+    let snapshotRate = new InterestRate(snapshotRateId);
+    snapshotRate.side = rate.side;
+    snapshotRate.type = rate.type;
+    snapshotRate.rate = rate.rate;
+    snapshotRate.save();
+    snapshotRates.push(snapshotRateId);
+  }
+  return snapshotRates;
+}
+
+export function setMarketVSTDebt(
+  event: ethereum.Event,
+  asset: Address,
+  debtVST: BigInt
+): void {
+  const debtUSD = bigIntToBigDecimal(debtVST);
+  const market = getOrCreateMarket(asset);
+  const debtUSDChange = debtUSD.minus(market.totalBorrowBalanceUSD);
+  const debtVSTChange = BigInt.fromString(
+    debtUSDChange.times(exponentToBigDecimal()).toString().split(".")[0]
+  );
+  market.totalBorrowBalanceUSD = debtUSD;
+  market.save();
+
+  getOrCreateMarketSnapshot(event, market);
+  getOrCreateMarketHourlySnapshot(event, market);
+  updateProtocolBorrowBalance(event, debtUSDChange, debtVSTChange);
+}
+
+export function setMarketAssetBalance(
+  event: ethereum.Event,
+  asset: Address,
+  balanceAsset: BigInt
+): void {
+  const balanceUSD = bigIntToBigDecimal(balanceAsset).times(
+    getCurrentAssetPrice(asset)
+  );
+  const market = getOrCreateMarket(asset);
+  const netChangeUSD = balanceUSD.minus(market.totalValueLockedUSD);
+  market.totalValueLockedUSD = balanceUSD;
+  market.totalDepositBalanceUSD = balanceUSD;
+  market.inputToken = asset.toHexString();
+  market.inputTokenBalance = balanceAsset;
+  market.inputTokenPriceUSD = getCurrentAssetPrice(asset);
+  market.save();
+
+  getOrCreateMarketSnapshot(event, market);
+  getOrCreateMarketHourlySnapshot(event, market);
+  updateProtocolUSDLocked(event, netChangeUSD);
+}
+
+export function addMarketDepositVolume(
```

Comment:
is there another reason to have these explicit functions, aside from being able to call them from the event handlers? To me they seem unnecessary, and you could just call `addMarketVolume()` from event.ts

Reply:
The original implementation have different details functions for borrow.repay and so on. After TrueFi subgraph review, I merged their details into one concrete function. Just remove these middle layer function in latest commit at 6fd7ecf7a0f6bb9ed6d41b5e08b7f9206f0fbb19